src/transform/src/attribute/cardinality.rs

```diff
@@ -0,0 +1,646 @@
+                continue;
+            }
+
+            // some unique columns in this equivalence
+            for idx in unique_sources {
+                // when joining R and S on R.x = S.x, if R.x is unique and S.x is not, we're bounded above by the cardinality of S
+                inputs[idx] = SymExp::f64(1.0);
+            }
+        }
+
+        SymbolicExpression::product(inputs)
+    }
+
+    fn reduce(
+        &self,
+        group_key: &Vec<MirScalarExpr>,
+        expected_group_size: &Option<u64>,
+        input: &SymExp,
+    ) -> SymExp {
+        // TODO(mgree): if no `group_key` is present, we can do way better
+
+        if let Some(group_size) = expected_group_size {
+            input / f64::cast_lossy(*group_size)
+        } else if group_key.is_empty() {
+            SymExp::from(1)
+        } else {
+            // in the worst case, every row is its own group
+            input.clone()
+        }
+    }
+
+    fn topk(
+        &self,
+        group_key: &Vec<usize>,
+        limit: &Option<usize>,
+        expected_group_size: &Option<u64>,
+        input: &SymExp,
+    ) -> SymExp {
+        let k = limit.unwrap_or(1);
+
+        if let Some(group_size) = expected_group_size {
+            input * (f64::cast_lossy(k) / f64::cast_lossy(*group_size))
+        } else if group_key.is_empty() {
+            SymExp::from(k)
+        } else {
+            // in the worst case, every row is its own group
+            input.clone()
+        }
+    }
+
+    fn threshold(&self, input: &SymExp) -> SymExp {
+        // worst case scaling factor is 1
+        input.clone()
+    }
+}
+
+impl Attribute for Cardinality {
+    type Value = SymExp;
+
+    fn derive(&mut self, expr: &MirRelationExpr, deps: &DerivedAttributes) {
+        use MirRelationExpr::*;
+        let n = self.results.len();
+
+        match expr {
+            Constant { rows, .. } => self
+                .results
+                .push(SymExp::from(rows.as_ref().map_or_else(|_| 0, |v| v.len()))),
+            Get { id, .. } => match id {
+                Id::Local(id) => match self.env.get(id) {
+                    Some(value) => self.results.push(value.clone()),
+                    None => {
+                        // TODO(mgree) when will we meet an unbound local?
+                        self.results
+                            .push(SymExp::symbolic(FactorizerVariable::Unknown));
+                    }
+                },
+                Id::Global(id) => self
+                    .results
+                    .push(SymbolicExpression::symbolic(FactorizerVariable::Id(*id))),
+            },
+            Let { .. } | Project { .. } | Map { .. } | ArrangeBy { .. } | Negate { .. } => {
+                let input = self.results[n - 1].clone();
+                self.results.push(input);
+            }
+            LetRec { .. } => unimplemented!("cardinality analysis does not yet support recursion"),
```

Comment:
Maybe use `FactorizerVariable::Unknown` instead of `unimplemented!`, which would panic. Or is there a mechanism somewhere that ensures that we don't call cardinality estimation on anything involving `LetRec`?